pkg/service/service.go

```diff
@@ -0,0 +1,117 @@
+package service
+
+import (
+	"github.com/readium/go-toolkit/pkg/fetcher"
+	"github.com/readium/go-toolkit/pkg/manifest"
+)
+
+// Base interface to be implemented by all publication services.
+type Service interface {
+	Links() []manifest.Link                          // Links to be added to the publication
+	Get(link manifest.Link) (fetcher.Resource, bool) // A service can return a Resource that supplements, replaces or compensates for other links
+	Close()                                          // Closes any opened file handles, removes temporary files, etc.
+}
+
+// Container for the context from which a service is created.
+type Context struct {
+	Manifest manifest.Manifest
+	Fetcher  fetcher.Fetcher
+}
+
+func NewContext(manifest manifest.Manifest, fetcher fetcher.Fetcher) Context {
+	return Context{
+		Manifest: manifest,
+		Fetcher:  fetcher,
+	}
+}
+
+type ServiceFactory func(context Context) Service
+
+// Builds a list of [Service] from a collection of service factories.
+// Provides helpers to manipulate the list of services of a [pub.Publication].
+type ServicesBuilder struct {
+	serviceFactories map[string]ServiceFactory
+}
+
+func NewBuilder(
+	contentProtection ServiceFactory,
+	cover ServiceFactory,
+	locator ServiceFactory,
+	positions ServiceFactory,
+	search ServiceFactory,
+) *ServicesBuilder {
```

Comment:
I feel like this API, at least in Go, will be pretty error prone: easy to invert two parameters as there are of the same type and we will add more services from time to time which means duplicating this API to keep backward compatibility.

In Swift/Kotlin we can/must expand the parameter names which alleviates this issue.

Maybe we can just add a `NewBuilder()` that takes a single `map[string]ServiceFactory`?

Reply:
Agreed, it looks much better